In [1]:
import numpy as np

In [19]:
arr1 = np.array([6, 2, -7, 2, 8, -2, 1])   #数组是连续储存的，并且是同一类数据类型，这有异于list
arr1

array([ 6,  2, -7,  2,  8, -2,  1])

In [14]:
print (arr1)

[ 6  2 -7  2  8 -2  1]


In [20]:
arr2 = np.array([[1,3,5],[2,4,6]])
arr2

array([[1, 3, 5],
       [2, 4, 6]])

In [16]:
arr2

array([[1, 3, 5],
       [2, 4, 6]])

$$分别打印arr1 和 arr2 的维度，形状，元素数量，元素类型$$

In [17]:
np.size(arr1)

7

In [30]:
print (arr1.ndim)
print (arr2.ndim)
print (arr1.shape)
print (arr2.shape)
print (arr1.size)
print (arr2.size)
print (arr1.dtype)
print (arr2.dtype)

1
2
(7,)
(2, 3)
7
6
int64
int64


In [31]:
arr_all_0 = np.zeros (6)  #六个0浮点数
arr_all_0 

array([0., 0., 0., 0., 0., 0.])

In [32]:
arr_all_1 = np.ones (6)  #六个1浮点数
arr_all_1 

array([1., 1., 1., 1., 1., 1.])

In [38]:
arr_even = np.arange (10, 21, 2) 
arr_even

array([10, 12, 14, 16, 18, 20])

In [41]:
arr_even.sum()

np.int64(90)

In [46]:
arr_even[(arr_even > 11) & (arr_even < 19)]

array([12, 14, 16, 18])

arr_even.sort()  #不改动 原arr_even，只是排序
np.sort (arr_even) # 改动


In [48]:
print (arr_even[2])
print (arr_even[-4]) #倒着数

14
14


In [50]:
arr_even [1:4] #切片功能 第一个为起始索引被包括，最后一个不被包括

array([12, 14, 16])

In [52]:
arr3 = np.array([1,2,5,2,-7,34,9])

In [55]:
arr4 = np.array([4,1,5,49,7,3,12])

In [56]:
arr3 + arr4 # 数组可以直接相加 前提是长度一致

for e1 in arr3:
    for e2 in arr4:
        e1 + e2  #如果是list 要写for loops进行相加

array([ 5,  3, 10, 51,  0, 37, 21])

In [57]:
arr3 * -5 #相乘

array([  -5,  -10,  -25,  -10,   35, -170,  -45])

In [58]:
arr3 < 1 #输出boolean expressions

array([False, False, False, False,  True, False, False])

In [59]:
arr3 [arr3 < 3] #这里才会输出具体数值

array([ 1,  2,  2, -7])

In [ ]:
arr3 [(arr3 <1)  ]    # and&  or|   not~

In [1]:
import pandas as pd #相当于导入后直接赋值给pd，后续输入就可以只写pd

In [3]:
s1 = pd.Series([4,12,2,-9,45,34])
s1 #会给出索引值（第一排）

0     4
1    12
2     2
3    -9
4    45
5    34
dtype: int64

In [5]:
s1.values #获得里面数值

array([ 4, 12,  2, -9, 45, 34])

In [6]:
s1.index # index属性 获得索引值

RangeIndex(start=0, stop=6, step=1)

In [8]:
s1 = pd.Series([4,12,2,-9,45,34], index=["a", "d", "h", "w", "t", "g"]) # pandas的特别之处，可以修改索引值即标签索引, 
s1

a     4
d    12
h     2
w    -9
t    45
g    34
dtype: int64

In [10]:
s1 ["a":"w"] # 切片时，索引要记得加上 “” ，最后一个包括， 位置索引最后一个不包括。如果标签索引是数字 比如1 4 6等，切片会以原始的位置进行索引，如果要用数字标签索引pandas加入了loc，列如s1.loc[1:4],  还有iloc

a     4
d    12
h     2
w    -9
dtype: int64

In [11]:
s1 [["a", "t","g"]] #索引指定需要的数值，s1 [[1,3,6]]位置索引也是如此，但是记得是[[]]

a     4
t    45
g    34
dtype: int64

创建series的另一种方式：加入dict字典

In [14]:
s2 = pd.Series ([4.1, 2.2, 5.3, 3.7, 6.8], index=["青菜", "白萝卜", "西红柿", "土豆", "黄瓜"])

In [15]:
s2 = pd.Series ({"青菜":4.1, "白萝卜":2.2, "西红柿":5.3, "土豆":3.7, "黄瓜":6.8}) #加入dict

In [16]:
"青菜" in s2 #返回一个boolean expression提示是否存在

True

In [17]:
s2 ["青菜"]

np.float64(4.1)

In [18]:
s2.loc["青菜"]

np.float64(4.1)

In [19]:
s2.iloc[0]

np.float64(4.1)

In [20]:
s2 [(s2>5) | (s2<3)] #根据条件需要筛选series元素

白萝卜    2.2
西红柿    5.3
黄瓜     6.8
dtype: float64

In [22]:
s5 = pd.Series([1,2,3,4,5])

In [26]:
s5 = pd.Series([1,2,3,4,5],index = [5,4,3,2,1])

In [29]:
s5

5    1
4    2
3    3
2    4
1    5
dtype: int64

In [32]:
s5.iloc[1]

np.int64(2)

In [34]:
s5.loc[5:3]

5    1
4    2
3    3
dtype: int64

In [38]:
s6 = pd.Series({"xiao li":82, "xiao chen":93, "xiao zhang":91, "xiao zeng":78, "xiao deng":68})
s6

xiao li       82
xiao chen     93
xiao zhang    91
xiao zeng     78
xiao deng     68
dtype: int64

In [41]:
s6["xiao zhang"] = 95  # 或者使用s3.loc ["xiao zhang"] = 95
s6

xiao li       82
xiao chen     93
xiao zhang    95
xiao zeng     78
xiao deng     68
dtype: int64

In [44]:
s6 >80

xiao li        True
xiao chen      True
xiao zhang     True
xiao zeng     False
xiao deng     False
dtype: bool

In [46]:
s6 [(s6 >80) & (s6 < 90)]

xiao li    82
dtype: int64

In [3]:
import pandas as pd

In [4]:
s9 = pd.Series ([92, 67, 70, 88, 76], index = ["ming", "hong", "jie", "li", "hua"])
s8 = pd.Series ([95, 85, 60, 79, 76], index = ["ming", "jie", "yu", "juan", "tong"])


In [5]:
s9

ming    92
hong    67
jie     70
li      88
hua     76
dtype: int64

In [6]:
s8

ming    95
jie     85
yu      60
juan    79
tong    76
dtype: int64

In [7]:
s9 + s8

hong      NaN
hua       NaN
jie     155.0
juan      NaN
li        NaN
ming    187.0
tong      NaN
yu        NaN
dtype: float64

In [8]:
s9.add(s8, fill_value=0)

hong     67.0
hua      76.0
jie     155.0
juan     79.0
li       88.0
ming    187.0
tong     76.0
yu       60.0
dtype: float64

In [13]:
s9.describe()

count     5.000000
mean     78.600000
std      10.990905
min      67.000000
25%      70.000000
50%      76.000000
75%      88.000000
max      92.000000
dtype: float64

In [16]:
s9 = s9 + 5 #需要再次赋值才会改变s9

In [20]:
def get_rank (scores):  #自定义函数，给分数排等级
    if scores >= 90:
        return "A"
    elif scores >=80:
        return "B"
    elif scores >=70:
        return "C"
    elif scores >= 60:
        return "D"
    else:
        return "fail"
get_rank(59)



'fail'

In [19]:
s9.apply (get_rank)

ming    A
hong    C
jie     B
li      A
hua     B
dtype: object

In [26]:
name = pd.Series(["chen", "li", "wang", "zha", "zao", "zio"], index=["001", "002", "003", "004", "005", "006"])
gender = pd.Series(["fm", "fm", "m", "m", "fm","fm"], index=["006","005", "004", "003", "002", "001"])
height = pd.Series ([172.5, 168.0, 178.2, 181.3, 161.7], index = ["001", "002", "003", "004", "005"])

In [31]:
students = pd.DataFrame({"姓名":name, "性别": gender, "身高": height}) 
students

姓名  性别     身高
001  chen  fm  172.5
002    li  fm  168.0
003  wang   m  178.2
004   zha   m  181.3
005   zao  fm  161.7
006   zio  fm    NaN

In [33]:
print(students.index)
print(students.columns)

Index(['001', '002', '003', '004', '005', '006'], dtype='object')
Index(['姓名', '性别', '身高'], dtype='object')


In [34]:
students.身高 # 或者 students["身高"]

001    172.5
002    168.0
003    178.2
004    181.3
005    161.7
006      NaN
Name: 身高, dtype: float64

In [36]:
students.T # transposition 

001    002    003    004    005  006
姓名   chen     li   wang    zha    zao  zio
性别     fm     fm      m      m     fm   fm
身高  172.5  168.0  178.2  181.3  161.7  NaN

In [39]:
students[["姓名","身高"]]

姓名     身高
001  chen  172.5
002    li  168.0
003  wang  178.2
004   zha  181.3
005   zao  161.7
006   zio    NaN

In [43]:
students.loc["003"]

姓名     wang
性别        m
身高    178.2
Name: 003, dtype: object

In [45]:
students["003":"005"]

姓名  性别     身高
003  wang   m  178.2
004   zha   m  181.3
005   zao  fm  161.7

In [46]:
students.loc ["005", "身高"]

np.float64(161.7)

In [47]:
students.loc [["003","005"], ["姓名","身高"]]

姓名     身高
003  wang  178.2
005   zao  161.7

In [48]:
students.loc[["003","005"], :]

姓名  性别     身高
003  wang   m  178.2
005   zao  fm  161.7

In [50]:
students ["身高"] > 165

001     True
002     True
003     True
004     True
005    False
006    False
Name: 身高, dtype: bool

In [51]:
students [(students ["身高"] > 165) & (students ["性别"] == "fm")]

姓名  性别     身高
001  chen  fm  172.5
002    li  fm  168.0

In [52]:
students.head() # ()默认前五行 可以自行定义

姓名  性别     身高
001  chen  fm  172.5
002    li  fm  168.0
003  wang   m  178.2
004   zha   m  181.3
005   zao  fm  161.7

In [ ]:

import pandas as pd

In [ ]:
student_data = {"x":{"Y":"Z", "A","B"}} # 创建dataframe ， dict 里面是行，外面的x对于列名
students = pd.DataFrame (student_data)

In [ ]:
students["考试4"] = [12,12,12,23,34,12,34,65,87] #在dataframe里面添加一列

In [ ]:
students.loc[007] = ["小杨"，79，80，82，67，87] # 添加一行，007为索引

In [ ]:
students.drop(["006","007"], axis = 0)  #删除 006 007行，axis默认为0可以不写（axis为0是对列进行操作，为1是对行进行操作），drop只删除，不对原students进行修改

In [ ]:
students.drop (["考试2","考试4"], axis = 1) 

In [ ]:
T = pd.Series({"考试1":90, "考试2":89,"考试3":78,"考试3":88, "考试4",67}) #创建series

In [ ]:
T + students[["考试1", "考试2", "考试3", "考试4"]] # T 与students相加，可以特意把students里的像对应的值提取出，进行单独相加

In [ ]:
students["考试4"] = students["考试4"] + 10  # 考试4的值全部加10，然后更新原students里的考试4那一列 ，不能一直点击输出，会一直加10，如果发现多点击了，可以在最上方菜单栏点击restart的标志就是重新输出第一次运行后结果.

In [ ]:
如果第三方库里面的函数方法无法做到你需要的数据处理结果，可以加接收函数 .apply 函数，比如 s.apply 和 df1.apply

In [ ]:
describe方法 df1.describe 就是各列的统计信息 而且只计算有数字的列，如果有字符会自动忽略

In [20]:
import pandas as pd
import numpy  as np

In [3]:
students = pd.DataFrame({"001":{"name":"XC", "score1":85, "score2":95,"score3":92}, "002":{"name":"XL", "score1":91, "score2":92,"score3":94}, "003":{"name":"XW", "score1":86, "score2":81,"score3":89}, "004":{"name":"XZ", "score1":79, "score2":89,"score3":95}, "005":{"name":"XZO", "score1":96, "score2":91,"score3":91}, "006":{"name":"XZU", "score1":81, "score2":89,"score3":92}})
students

001 002 003 004  005  006
name    XC  XL  XW  XZ  XZO  XZU
score1  85  91  86  79   96   81
score2  95  92  81  89   91   89
score3  92  94  89  95   91   92

In [4]:
students = pd.DataFrame({"001":{"name":"XC", "score1":85, "score2":95,"score3":92}, "002":{"name":"XL", "score1":91, "score2":92,"score3":94}, "003":{"name":"XW", "score1":86, "score2":81,"score3":89}, "004":{"name":"XZ", "score1":79, "score2":89,"score3":95}, "005":{"name":"XZO", "score1":96, "score2":91,"score3":91}, "006":{"name":"XZU", "score1":81, "score2":89,"score3":92}}).T
students

name score1 score2 score3
001   XC     85     95     92
002   XL     91     92     94
003   XW     86     81     89
004   XZ     79     89     95
005  XZO     96     91     91
006  XZU     81     89     92

In [5]:
students.loc[:, "score1":"score3"].mean() # 第一个:是包括001到006的切片，loc是包含最后一个，iloc则不包含

score1    86.333333
score2         89.5
score3    92.166667
dtype: object

In [9]:
#上面默认给出了每列的平均值 （axis=0），但我们想要的是每一行的平均值

In [7]:
average = students.loc[:, "score1":"score3"].mean(axis=1)

In [17]:
name1 = students["name"]
name1

001     XC
002     XL
003     XW
004     XZ
005    XZO
006    XZU
Name: name, dtype: object

In [19]:
students_average1 = pd.DataFrame({"姓名":name1, "平均分":average})
students_average1 

姓名        平均分
001   XC  90.666667
002   XL  92.333333
003   XW  85.333333
004   XZ  87.666667
005  XZO  92.666667
006  XZU  87.333333

In [ ]:
# 输出每次考试第二高的分数组成的Series， 可以尝试结合匿名函数 （适用于逻辑较为简单的，逻辑复杂则使用def）

In [21]:
students.loc [:,"score1":"score3"].apply(lambda x: np.sort(x)[-2])  # x 是自定义的 ,利用np库sort函数进行排序在返回给 X，倒数第二个值即为第二高的值（因为先排序了）

score1    91
score2    92
score3    94
dtype: int64

In [29]:
def get_grade(sco):
    if sco >= 95:
        return "A+"
    elif sco >= 90:
        return "A"
    elif sco >=85:
        return "B+"
    elif sco >= 80:
        return "B"
    elif sco >= 75:
        return "C+"
    else:
        return "C"

In [ ]:
students.loc[:,"score1":"score3"] #先提取出来 数字

In [30]:
students.loc[:,"score1":"score3"].applymap(get_grade) # 再进行函数执行

C:\Users\ooazh\AppData\Local\Temp\ipykernel_39400\4283995628.py:1: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  students.loc[:,"score1":"score3"].applymap(get_grade) # 再进行函数执行


score1 score2 score3
001     B+     A+      A
002      A      A      A
003     B+      B     B+
004     C+     B+     A+
005     A+      A      A
006      B     B+      A

In [32]:
students.describe() # 当成字符串执行了 （string是object的子类）

name  score1  score2  score3
count     6       6       6       6
unique    6       6       5       5
top      XC      85      89      92
freq      1       1       2       2

In [ ]:
students["score1"].astype("int") # 所以 转换成int类型

In [34]:
students["score1"] = students["score1"].astype("int")
students["score2"] = students["score2"].astype("int")
students["score3"] = students["score3"].astype("int")
students.describe() 

score1     score2     score3
count   6.000000   6.000000   6.000000
mean   86.333333  89.500000  92.166667
std     6.314006   4.722288   2.136976
min    79.000000  81.000000  89.000000
25%    82.000000  89.000000  91.250000
50%    85.500000  90.000000  92.000000
75%    89.750000  91.750000  93.500000
max    96.000000  95.000000  95.000000

name  score1  score2  score3
count     6       6       6       6
unique    6       6       5       5
top      XC      85      89      92
freq      1       1       2       2

**1 获取数据，分内部和外部获取，公司内部获取可能需要给程序员提出需求/或者自行写查询语句来提取。而公开数据的获取方法有直接下载/爬虫/API等**

爬虫：1.发送请求后并且只能获取网页源代码，第二就必须解析网页源代码内容，才能提取出想要的内容。
而API：根据合约要求直接获得需要的数据内容，api一般会以JSON格式返回（纯文本即无字号颜色等特征，程序员喜欢的格式），数据结构类型包括：JSON对象
{...包括string ，int，bool，null， float，
 还可以加入数组[...] 

}和数组[]与python字典和列表类似

当获取了JSON文件需要解析+转换成python可读的, panda库里有pd.read_json("/输入文件路径/")， 键名key对应dataframe的列名，而value值对应列名下面的数据

文件格式：比如data.txt，后缀如果修改，比如data.mp3 那么电脑就会以音乐软件取打开此文件，但实际内容不会改变。

CSV格式（也是纯文本格式，以comma逗号分隔和规整的二维结构为规则，体积小易读取，与dataframe结构类型，而json层层嵌套很难看出对应表格什么样）则是数据分析师最爱，

**2 解析数据**
当文件数据量大时，如果excel打开可能会卡住，excel展示额外格式，加载额外功能等等。所以可以用代码编辑器vs code等和纯文本编辑器打开。当还可以用代码读取转换成dataframe，然后用pandas库里面的方法，自定义指定查看，而不用全部海量加载出来。所以用代码处理数据比用软件要方便又快.。


如果原始文件没有列名，也就是第一行就是数据了，解析时加header = None即不要把第一行当成列名，dataframe就会把第一行当成第一条数据
pd.read_csv("/输入文件路径/", header = None)，不然csv默认第一行就是列名
但是不会把第一列默认解析为索引，而是有自带默认的位置索引，如果需要指定某一列为标签索引，可以传入index_col这个可选参数 
pd.read_json("/输入文件路径/"， index_col = "xxx") 指定了xxx列为标签索引，进而考索引提取目标数据


当列数太多，dataframe会省略到20列，如果需要多查看，可以用pandas的set_option 设置需要看到的列数的上限  pd.set_option ("display.max_colums", 150)
然后有些值长度太长，没有展示全 默认为50，可以 pd.set_option ("display.max_colwidth", 500) 表示每一列里的值最多包含几个字符，例子中500即为上限。

df1.head()  df1.tail() df1.info()得到概况信息几行几列数据类型等

**3 评估数据（清洗数据，整洁度和干净度的评估）**
标准：结构和内容
结构上来讲, 处理前的数据为乱数据 与之对应为整洁数据即根据**埃德加·科德的第三范式 包括三个特点 1每列是一个变量(如都是身高的变量) 2每行是一个观察值(某一个人或者物信息) 3每个单元格是一个值**
（关于变量的理解可再次翻看视频！！！经典）很多方便人类理解的excel表格是属于不整洁数据的

从内容上讲，处理前的数据为脏数据（内容方面需要清洗），反之为干净数据。脏数据包括 **1丢失值NaN 2重复数据 3不一致数据 4无效或者错误数据**

整洁度 结构很直观的容易被发现纠正，但是干净度 里面的内容太多不易check
假如scores是数据结构dataframe，scores.isnull().sum() 即可查找空值的数量 也可以提取出某一列的空缺值数量 scores["考试3"].isnull.sum()， 如果要查看空缺值的行 scores[scores["考试3"].isnull()],  评估重复值：查看重复数据用duplicated， 评估不一致数据（比如一班和1班）：value_counts()，  评估无效/错误值（比如 离群值）：排序sort函数或者describe也可以

**4 清洗数据**
第一步先观察 结构：如果不符合第三范式，需要进行转置、拆分变量（62课可再次查看）
第二步 内容：对4个特征进行评估，缺失值根据需求处理 可以不处理/可以人工填入/可以删除行/可以平均值填充等；重复数据找出后删除即可；不一致数据需要进行统一；无效数据删除或者替换。
具体函数：重命名索引名和列名： rename函数，但只是生效，不会修改原数据，需要重新赋值新变量/或者加入inpalce=Ture就可以原地修改
大写行名 df2.rename(columns= str.upper), 设置某一列为索引df2.set_index("输入所需要的索引列名")； df2.sort_index(inplace=ture)表示排序索引(因为索引顺序乱)，inpalce指的原地改变原函数；
对列进行拆分，str.split("/", expand = True)表示将原本在一列里的两个变量拆分成两个并分布两列， 用drop函数删除原本的列； 合并列 str.cat；转换成长数据melt函数（具体看64课），，还有explode函数。
（series方法）删除某一行 df.drop(2) 表示删除第三行，axis默认等于0就是读取行，等于1就是执行列
对于缺失值需要定位它 series方法 df3.loc["001":"009", "日期"] = "2029-01-09" 提取后在替换；或者df3["B"].fillna(df["B"].mean()) 表示找到所有空值并替换成平均值
（*series方法就是基于索引加列名而执行 对某一行或者列进行修改，而dataframe的方法可以基于键值对进行修改替换*）
dataframe结构 df5.dropna() 表示删除空值所在的行 ， df.dropna(subset = ["工资"]) 即查看工资这一列为空缺值的位置，并删除所在行. df.dropna(axis=1) 每一列有空值就删.

删除重复值 Series调用  df6["姓名"].drop_duplicates()   DateFrame调用 df6.drop_duplicates

对值类型进行转换 str int float bool list..  还有一种数据类型category 分类数据： 数据可以被分为分类数据（男女/ 金银铜牌...）和数值数据（具体数值）。
如果astype转换成category时 可以节约内存空间/并且有利于pandas自动选用合适的统计方法可视化。

保存清洗后的数据：df1.to_csv("xxx.csv")

In [ ]:
# 按DD删除  按m转换成markdown  按b 扩列一行  shift+enter是执行 

## 读取数据 reading data

In [3]:
import pandas as pd

In [7]:
original_data = pd.read_csv("C:/Users/ooazh/AppData/Local/Temp/baiduyunguanjia/onlinedit/cache/230bdfd15d0a26448875b7d26790ad13/e_commerce.csv")

In [8]:
original_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 评估数据 evaluating date

在这一步，我将对上一部分建立的original data 进行评估

主要分为：1整洁度 2干净度

### 评估数据整洁度

In [14]:
original_data.sample(10)  # sample some data randomly

InvoiceNo StockCode                       Description  Quantity  \
147456    549118     22550                  HOLIDAY FUN LUDO         4   
410477    572103     22440     BALLOON WATER BOMB PACK OF 35        10   
294760    562717    46776F  WOVEN ROSE GARDEN CUSHION COVER          1   
236225    557745     22031  BOTANICAL LAVENDER BIRTHDAY CARD        12   
523088    580501     21411       GINGHAM HEART  DOORSTOP RED         2   
96517     544569     22907  PACK OF 20 NAPKINS PANTRY DESIGN         2   
180568    552327     22360   GLASS JAR ENGLISH CONFECTIONERY        12   
225208    556636     84600        NEW BAROQUE JEWELLERY BOX          1   
328909    565838     23092   LARGE ANTIQUE WHITE PHOTO FRAME        12   
60666     541422    84968A    SET OF 16 VINTAGE ROSE CUTLERY         2   

             InvoiceDate  UnitPrice  CustomerID         Country  
147456    4/6/2011 13:18       3.75     14911.0            EIRE  
410477  10/20/2011 15:57       0.42     17571.0  United Kingdom  
294760    8/8/2011 15:42       4.13         NaN  United Kingdom  
236225   6/22/2011 12:16       0.42     15750.0  United Kingdom  
523088   12/4/2011 13:00       1.95     14546.0  United Kingdom  
96517    2/21/2011 12:58       0.85     16980.0  United Kingdom  
180568    5/9/2011 10:03       2.95     17007.0  United Kingdom  
225208   6/13/2011 15:30      15.79         NaN  United Kingdom  
328909    9/7/2011 11:43       6.65     15071.0  United Kingdom  
60666    1/17/2011 17:48       8.29         NaN  United Kingdom

从随机的10行数据来看，数据符合第三范式

### 评估数据干净度

In [15]:
original_data.info() # check where i can see which variable， and 每个变量里有多少非空值，以及数据类型是什么

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


从结果输出来看，数据共有541909条观察值，而变量description, customerID存在缺失值。

此外，InvoiceData 的数据类型应该是日期，CustomerID 应该是string，需要进行格式转换。

#### 评估缺失数据

In [20]:
original_data[original_data["Description"].isnull()] # [original_data["Description"].isnull() 这一部分以description是否为null来筛选数据，然后获得其结果布尔值Series，然后把bool值作为索引extract “Description”的空值

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

In [21]:
original_data[(original_data["Description"].isnull()) & (original_data["UnitPrice"] != 0)] # 验证 des（商品名）的值是否为0且uni（单价）的值不等于0

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

产品名和单价都缺失/无效，我们认为数据无法提供我们分析目的的有效含义，后续可以删除。

In [23]:
original_data[original_data["CustomerID"].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

CustomerID	表示客户编号，不是分析畅销商品的必要变量，有些CustomerID 缺失的交易数据仍然有效，因此保留此变量为空的观察值。

#### 评估重复数据

对唯一标识符进行评估，即不能重复的数据。InvoiceNo	StockCode	CustomerID	都是唯一标识符，但这里的情况可以出现重复值，所以无需数据评估。

#### 评估不一致数据

不一致数据可能存在于Country，我们需要查看是否存在多个不同值指代同一个国家的情况

In [26]:
original_data["Country"].value_counts() #extract country的series，发现英国和美国有多个指代对象

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


#### 评估无效或者错误数据

In [27]:
original_data.describe() #基础DataFrame的describe方法，对数值统计信息进行快速了解

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

从输出结果来看，CustomerID不应该被计算，因为其代表编号，后续需要进行数据类型转换成string。     
Quantit 和UnitPrice存在负值，会对后续数值分析产生影响，因此，我们需要先筛选出其为负数的观察值，进一步评估其含义。	

In [29]:
original_data[original_data["Quantity"] < 0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

Quantity 为负数的发票开头都是C  （这里有个背景信息是，c开头发票为被取消的订单）

In [33]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C")]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

这里结果得出，存在qua为负值而发票不以c开头的编号，并且单价都为0

In [34]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C") & (original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

说明，quan为负值要么发票开头为C 要么商品单价为0，这些数据都不是有效成交数据，不贡献销售，而且会影响我的数据分析， 我们的分析目的是找出畅销商品，后续可以删除

UnitPrice也有是负数的

In [36]:
original_data[original_data["UnitPrice"] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

UnitPrice为负数的都是Adjust bad debt坏账，不属于实际商品数据，后续删除。

## 清洗数据

基础评估后待办的分析任务有：
1 把
2
3
...

In [38]:
cleaned_data = original_data.copy() # 将原始数据备份 一份给要清洗的数据， 存在两份数据
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [40]:
cleaned_data["InvoiceDate"] = pd.to_datetime(cleaned_data["InvoiceDate"]) #将InvoiceDate数据类型into为 datetime数据类型(不常见数据类型) by pandas库里的to_datetime函数
cleaned_data["InvoiceDate"] 

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

In [41]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].astype(str) #CustomerID转换成string类型	
cleaned_data["CustomerID"]

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

In [43]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].str.slice(0,-2) # 删除其值后面的.0, 加.str表示可以调用能操作字符串的函数。（series）
cleaned_data["CustomerID"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

In [45]:
cleaned_data.dropna(subset = ["Description"], inplace=True) # 对dataframe进行空值删除，删除哪里？就是Description变量缺失的观察值; inplace=True原地改变可以不用进行赋值

In [46]:
cleaned_data["Description"].isnull().sum() # 对Des的series进行bool运算获得ture和false，求总和就是ture的数量（false=0， ture=1）

np.int64(0)

这里求和为0，则空值已被删除完毕。

In [48]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"USA":"United States"}) # 替换usa为united states 用series自带的replace函数

In [49]:
len(cleaned_data[cleaned_data["Country"] == "USA"]) #通过长度的方法 查看USA是否被替换完成，如果是0即替换完成。 (==为是否相等的boolean expression，只会返回ture/false)

0

In [50]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"UK":"United Kingdom" , "U.K.":"United Kingdom"})

In [54]:
len(cleaned_data [cleaned_data["Country"] == "UK"])

0

In [55]:
len(cleaned_data[cleaned_data["Country"] == "U.K."])

0

In [59]:
cleaned_data = cleaned_data[cleaned_data["Quantity"] >= 0] # 把quantity里的负值删除，但是没有一种根据条件删除数值的方法，所以提取 符合条件的数值 再重新赋值。

In [60]:
len(cleaned_data[cleaned_data["Quantity"] < 0])

0

In [62]:
cleaned_data = cleaned_data[cleaned_data["UnitPrice"] >= 0]

In [63]:
len(cleaned_data[cleaned_data["UnitPrice"] < 0])

0

## 保存清洗后的数据

In [64]:
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice CustomerID         Country  
0 2010-12-01 08:26:00       2.55      17850  United Kingdom  
1 2010-12-01 08:26:00       3.39      17850  United Kingdom  
2 2010-12-01 08:26:00       2.75      17850  United Kingdom  
3 2010-12-01 08:26:00       3.39      17850  United Kingdom  
4 2010-12-01 08:26:00       3.39      17850  United Kingdom

在保存时，index 1 2 3 ...都会被保存进去，所以如果不需要它，可以加入index=false

预将保存文件取名为e_commerce_cleaned

In [67]:
cleaned_data.to_csv("C:/Users/ooazh/AppData/Local/Temp/baiduyunguanjia/onlinedit/cache/230bdfd15d0a26448875b7d26790ad13/e_commerce_cleaned.csv", index=False)  # 可以指定位置保存

In [68]:
pd.read_csv("e_commerce_cleaned.csv").head()

C:\Users\ooazh\AppData\Local\Temp\ipykernel_28976\2686214642.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv("e_commerce_cleaned.csv").head()


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice CustomerID         Country  
0  2010-12-01 08:26:00       2.55      17850  United Kingdom  
1  2010-12-01 08:26:00       3.39      17850  United Kingdom  
2  2010-12-01 08:26:00       2.75      17850  United Kingdom  
3  2010-12-01 08:26:00       3.39      17850  United Kingdom  
4  2010-12-01 08:26:00       3.39      17850  United Kingdom